src/AstSemanticChecker.cpp

```diff
@@ -745,16 +745,73 @@ void AstSemanticChecker::checkRules(ErrorReport& report, const TypeEnvironment&
 
 // ----- types --------
 
+// check if a union contains a number primitive
+static bool unionContainsNumber(const AstProgram& program, const AstUnionType& type) {
+    for (const AstTypeIdentifier& sub : type.getTypes()) {
+        if (sub == "number") {
+            return true;
+        }
+        const AstType* subt = program.getType(sub);
```

Comment:
The variable names could be more informative `u` could be `union`, for example. Maybe `typeID` and `type` instead of `sub` and `subt` if that doesn't clash with existing symbols.

Reply:
`union` is a terrible name for a variable

Reply:
Fixed in e7403c697c16c269f500fcee3a5ccfa82a735d63